doc/dotNET SDK Proposal.md

```diff
@@ -0,0 +1,232 @@
+
+
+**.NET asynchrony does not create concurrency.** <a name="DotNetAsyncIsLogicallySingleThreaded"></a>
+Crucially, async/await-based asynchrony in .NET is _not_ implicitly multi-threaded. This means that invoking or awaiting [Task](https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/task-based-asynchronous-pattern-tap)-based async APIs does not result in logical multi-threading, except when a parallel execution thread is explicitly requested via one of the corresponding APIs. More on this topic is discussed in section [@ToDo](.).
+
+**Therefore, it is permitted to use async/await-based code in workflows. In fact, all Temporal interactions in .NET workflows are asynchronous and therefore all .NET workflows are required to be asynchronous.**
+
+**Alternative considerations.**
+It must be noted that other language SDKs take a variety of approaches to the multi-threading restrictions. This is required asynchronous and parallel programming is handled differently across runtimes. For example, the Java SDK permits "pseudo-parallelism": It permits workflows to invoke a subset of threading API. The Java SDK takes over the scheduling of such threads and ensures that such scheduling is deterministic.
+
+In .NET such approach is feasible, but we made an explicit decision not to provide such features in version 1. Two main reasons drove this decision:
+ 1. Because of the logically single-threaded nature of .NET async APIs, all workflow scenarios are better implemented using logically single threaded code. If, however, developers opted into _explicitly_ using multi-threaded parallelism (rather than asynchrony), then their code very likely depends on the parallelism to work as expected. A custom deterministic scheduler will invalidate developers' assumptions and lead to corruption. Such corruptions may be hard to diagnose and understand. Prohibiting multi-threading altogether leads developers into a pit of success.
+ 2. Adding features is easier than removing problematic features. We will listen to our user community and if, with time, we learn that there is an existing need to relax the single-threading restrictions in some particular way, we will - naturally - work to understand the underlying scenarios and to evolve the SDK accordingly. This is less risky than to provide features that are hard to use and can cause hard-to-diagnose reliability issues for some developers, and not being able to remove such features later, because other developers took a dependency on them.
+
+**Environment interactions:**
+
+**SDK offers deterministic alternatives to frequently used non-deterministic APIs.**
```

Comment:
Agree. We should touch on that in the discussion, even if we implement it later.

Reply:
:+1: The primary reason I bring it up now is because it can affect worker/workflow instantiation. For instance, if there is some kind of .Net sandbox construct that means they have to pass in workflow code instead of just class type, we can incorporate that. TS has to do the same.

I also just so happen to find this the most interesting part of any SDK development. I have found about a dozen ways to hack this in Go (and I know exactly how I'd do it w/ advanced JVM constructs).

Reply:
Added to next doc version.

✔️